aptos-move/framework/aptos-framework/sources/coin.move

```diff
@@ -138,6 +149,109 @@ module aptos_framework::coin {
+        coin_to_fungible_asset_map: SmartTable<TypeInfo, address>,
+    }
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    /// The paired coin type info stored in fungible asset medata object.
+    struct PairedCoinType has key {
+        type: TypeInfo,
+    }
+
+    inline fun borrow_conversion_map(): &CoinConversionMap {
+        if (!exists<CoinConversionMap>(@aptos_framework)) {
+            move_to(&create_signer::create_signer(@aptos_framework), CoinConversionMap {
+                coin_to_fungible_asset_map: smart_table::new<TypeInfo, address>(),
+            })
+        };
+        borrow_global<CoinConversionMap>(@aptos_framework)
+    }
+
+    inline fun borrow_conversion_map_mut(): &mut CoinConversionMap {
+        if (!exists<CoinConversionMap>(@aptos_framework)) {
+            move_to(&create_signer::create_signer(@aptos_framework), CoinConversionMap {
+                coin_to_fungible_asset_map: smart_table::new<TypeInfo, address>(),
+            })
+        };
+        borrow_global_mut<CoinConversionMap>(@aptos_framework)
+    }
+
+    /// Get the paired fungible asset metadata object of a coin type, create if not exist.
+    public fun paired_metadata<CoinType>(): Object<Metadata> acquires CoinConversionMap, CoinInfo {
+        let map = borrow_conversion_map_mut();
+        let type = type_info::type_of<CoinType>();
+        if (!smart_table::contains(&map.coin_to_fungible_asset_map, type)) {
+            let metadata_object_cref =
+                if (type_info::type_name<CoinType>() == string::utf8(b"0x1::aptos_coin::AptosCoin")) {
+                    object::create_object_at_address(@aptos_framework, @aptos_framework, false)
+                } else {
+                    let default_paired_metadata_address = object::create_object_address(&coin_address<CoinType>(), PAIRED_FUNGIBLE_ASSET_METADATA_SEED);
+                    if (!object::object_exists<Metadata>(default_paired_metadata_address)) {
+                        object::create_named_object(&create_signer::create_signer(coin_address<CoinType>()), PAIRED_FUNGIBLE_ASSET_METADATA_SEED)
+                    } else {
+                        object::create_sticky_object(coin_address<CoinType>())
+                    }
+                };
+            primary_fungible_store::create_primary_store_enabled_fungible_asset(
+                &metadata_object_cref,
+                option::map<u128, u128>(coin_supply<CoinType>(), |_| MAX_U128),
+                name<CoinType>(),
+                symbol<CoinType>(),
+                decimals<CoinType>(),
+                string::utf8(b""),
+                string::utf8(b""),
+            );
+            move_to(&object::generate_signer(&metadata_object_cref), PairedCoinType {type});
+            let metadata_addr = object::address_from_constructor_ref(&metadata_object_cref);
+            smart_table::add(&mut map.coin_to_fungible_asset_map, type, metadata_addr);
```

Comment:
The `smart_table::add` API has a DoS vector for overflowing a bucket to intentionally trigger `EEXCEED_MAX_BUCKET_SIZE`: basically just register a bunch of coin types to fill up the table until no one can use the table.

In a high-throughput chain with inexpensive txns like Aptos in practice it is worth it to register a bunch of bogus coins:

1. Take out a short position on `APT`
2. Register millions of coin types and invoke an FA transfer until hitting the error code described above
3. DoS `Coin` activity on the network
4. Close short position

How about replacing `smart_table` with the dumber but DoS resistance `table`?

Reply:
yes, I am not quite sure smart table is the right data structure to be used here. I think there are only hundreds of coins on chain and we will disable the new coin creation once we are going to do the migration. But to be careful, I think you are right. Let's move this to normal `table`.